core/trino-main/src/test/java/io/trino/execution/buffer/TestSpoolingOutputStats.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.buffer;
+
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertEquals;
+
+public class TestSpoolingOutputStats
+{
+    @Test
+    public void test()
+    {
+        int numberOfPartitions = 15;
+        SpoolingOutputStats spoolingOutputStats = new SpoolingOutputStats(numberOfPartitions);
+        assertThat(spoolingOutputStats.getFinalSnapshot()).isEmpty();
+        long[] expectedValues = new long[numberOfPartitions];
+        long value = 1;
+        for (int partition = 0; partition < numberOfPartitions; partition++) {
+            spoolingOutputStats.update(partition, value);
+            expectedValues[partition] = value;
+            value *= 31;
+        }
+        assertThat(spoolingOutputStats.getFinalSnapshot()).isEmpty();
+        spoolingOutputStats.finish();
+        assertThat(spoolingOutputStats.getFinalSnapshot()).isPresent();
+
+        // update is allowed to be called after finish, the invocation is ignored
+        spoolingOutputStats.update(0, value);
+        // finish is allowed to be called multiple times, the invocation is ignored
+        spoolingOutputStats.finish();
+        assertThat(spoolingOutputStats.getFinalSnapshot()).isPresent();
+
+        SpoolingOutputStats.Snapshot snapshot = spoolingOutputStats.getFinalSnapshot().orElseThrow();
+        assertEquals(snapshot.getPartitionSizeInBytes(0), 1);
+
+        for (int partition = 0; partition < numberOfPartitions; partition++) {
+            assertApproximatelyEquals(expectedValues[partition], snapshot.getPartitionSizeInBytes(partition));
+        }
+    }
+
+    private static void assertApproximatelyEquals(long expected, long actual)
+    {
+        // margin of error is expected to be not greater than 1%
```

Comment:
why 1%?

Reply:
Also I think you can just write:
```
assertThat(actual).isCloseTo(expected, withPercentage(1L));
```

Reply:
> why 1%?

There's a total of 23 precision bits in IEEE 754 single-precision 32-bit float (https://en.wikipedia.org/wiki/Single-precision_floating-point_format). The algorithm keeps the first 8 most significant bits of mantissa while discarding the remaining least significant 15 bits. 8 bits represent 256 unique values. The remaining least significant 15 bits represent a fraction of 1/256. Thus the precision loss should be less than 1/256 what is `<0.39%`.

Changed to 

> assertThat(snapshot.getPartitionSizeInBytes(partition)).isCloseTo(expectedValues[partition], withPercentage(EXPECTED_PRECISION_LOSS_IN_PERCENTS));

Where 

> double EXPECTED_PRECISION_LOSS_IN_PERCENTS = 100.0 / 256

And left a comment

Reply:
> assertThat(actual).isCloseTo(expected, withPercentage(1L));

Didn't know about this one, thanks for the suggestion